### PR TITLE
better scalar query

### PR DESCRIPTION
**Pull request by amosbird**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Category (leave one):
- Performance Improvement

Short description (up to few sentences):
Using column instead of AST to store scalar query results.

@yuzhichang @yingfeng

### CODE REVIEW

**Comment by yingfeng** [comment]
Awesome job!

**Comment by amosbird** [comment]
To-do ,  use generic ASTLiteral instead of ASTFunction

**Comment by amosbird** [comment]
perf test shows promising results

**Comment by yuzhichang** [comment]
Great job! Tested with product data, found no problem.

File: dbms/src/Client/Connection.cpp [hunk 1]
```diff
@@ -455,7 +456,10 @@ void Connection::sendData(const Block & block, const String & name)
         block_out = std::make_shared<NativeBlockOutputStream>(*maybe_compressed_out, server_revision, block.cloneEmpty());
     }
 
-    writeVarUInt(Protocol::Client::Data, *out);
+    if (scalar)
```
Comment (4ertus2) [inline dbms/src/Client/Connection.cpp]: Do we really need backward incompatible changes here?

Comment (amosbird) [reply dbms/src/Client/Connection.cpp]: It's compatible. Old servers just don't send any scalars

File: dbms/src/Interpreters/ExecuteScalarSubqueriesVisitor.cpp [hunk 2]
```diff
@@ -53,69 +56,95 @@ void ExecuteScalarSubqueriesMatcher::visit(ASTPtr & ast, Data & data)
...
-        exp_list->children.resize(columns);
-        for (size_t i = 0; i < columns; ++i)
-        {
-            exp_list->children[i] = addTypeConversionToAST(
-                std::make_unique<ASTLiteral>((*block.safeGetByPosition(i).column)[0]),
-                block.safeGetByPosition(i).type->getName());
-        }
+        auto func = makeASTFunction("__getScalar", std::make_shared<ASTLiteral>(scalar_query_hash_str));
+        func->alias = subquery.alias;
+        func->prefer_alias_to_column_name = subquery.prefer_alias_to_column_name;
```
Comment (4ertus2) [inline dbms/src/Interpreters/ExecuteScalarSubqueriesVisitor.cpp]: ```prefer_alias_to_column_name``` is a hack and deprecated for code that can be done without it. Can't we avoid it?

File: dbms/src/Client/Connection.cpp (hunk 1 above)
Comment (amosbird) [reply dbms/src/Client/Connection.cpp]: Hmm, what kind of backward incompatibility do you refer? Do you mean running a cluser with both old and new versions of clickhouse instances?

File: dbms/src/Interpreters/ExecuteScalarSubqueriesVisitor.cpp (hunk 2 above)
Comment (amosbird) [reply dbms/src/Interpreters/ExecuteScalarSubqueriesVisitor.cpp]: hmm, I do see some references to `prefer_alias_to_column_name`. I'm not familiar with that. Perhaps it deserves a separate pr to clean them all.

File: dbms/src/Client/Connection.cpp (hunk 1 above)
Comment (4ertus2) [reply dbms/src/Client/Connection.cpp]: Yes. It looks like we cannot upgrage or downgrade without downtime. Some installations has different update time for nodes with Distributed tables and their data sources. It's better to keep possibility to upgrade or downgrade them separately.

Comment (amosbird) [reply dbms/src/Client/Connection.cpp]: ok, I can add an scalar optimization option and default to false if that helps

**Comment by 4ertus2** [comment]
I've updated PR description. We use template strings in scrypts. PR should have "Category" and "Short description" unchanged.

**Comment by amosbird** [comment]
`enable_scalar_subquery_optimization` is set to `true` by default in order to get the perf test result. So there is a backward incompatibility and one might update the settings profile and set it to false so that a cluster installation can be upgraded without downtime.

**Comment by SaltTan** [comment]
If I set `enable_scalar_subquery_optimization` to anything I will get `Exception: Unknown setting enable_scalar_subquery_optimization` from older versions in distributed queries.

**Comment by den-crane** [comment]
> `enable_scalar_subquery_optimization` is set to `true`

@amosbird you made upgrade of a cluster is impossible.

**Comment by amosbird** [comment]
@SaltTan but `enable_scalar_subquery_optimization` only affects newer version of clickhouse. Older versions should behave as if it is set to `false`

@den-crane Hmm, did you try upgrading with that setting = `false` using config file?

**Comment by den-crane** [comment]
@amosbird   
= false  Yes, the upgraded server  starts to respond with errors `Exception: Unknown setting enable_scalar_subquery_optimization` because not upgraded shards know nothing about this setting.

**Comment by amosbird** [comment]
Ok, I'm not familiar with cluster upgrading, especially seemingless ones. If defaulting the setting to false is a proper fix, I'm fine with that. As I said,  "enable_scalar_subquery_optimization is set to true by default in order to get the perf test result."

**Pull request merged**